### PR TITLE
Sulu changes: Added possiblity to register custom adapters etc.

### PR DESCRIPTION
**Pull request by dantleech**
This PR allows custom adapters to be registered with the `cmf_routing_auto.adapter` tag, and for the adapter to be specified using the `cmf_routing_auto: { adapter: doctrine_phpcr_odm }` flag.

The default configuration is `doctrine_phpcr_odm`.

### CODE REVIEW

**Comment by dantleech** [comment]
Need to make a corresponding doc PR.

**Comment by dantleech** [comment]
Hmm, this conflicts with the `persistence` configuration..

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -26,7 +26,11 @@ public function getConfigTreeBuilder()
     {
         $treeBuilder = new TreeBuilder();
         $treeBuilder->root('cmf_routing_auto')
+            ->addDefaultsIfNotSet()
             ->children()
+                ->scalarNode('adapter')
+                    ->defaultValue('doctrine_phpcr_odm')
+                ->end()
```
Comment (WouterJ) [inline DependencyInjection/Configuration.php]: the CMF CS is to put this all on one line

File: DependencyInjection/Compiler/AutoRoutePass.php [hunk 2]
```diff
@@ -23,6 +23,52 @@ class AutoRoutePass implements CompilerPassInterface
 {
     public function process(ContainerBuilder $container)
     {
+        $this->registerServices($container);
+        $this->registerAdapter($container);
```
Comment (WouterJ) [inline DependencyInjection/Compiler/AutoRoutePass.php]: we should use 2 seperate compiler passes imo

File: DependencyInjection/Compiler/AutoRoutePass.php [hunk 3]
```diff
@@ -23,6 +23,52 @@ class AutoRoutePass implements CompilerPassInterface
...
+                throw new \InvalidArgumentException(sprintf(
+                    'No "name" specified for auto route adapter "%s"',
+                    $id
+                ));
+            }
+
+            $alias = $attributes[0]['alias'];
+            $adapterNames[] = $alias;
+            if ($adapter === $alias) {
+                $adapterId = $id;
```
Comment (WouterJ) [inline DependencyInjection/Compiler/AutoRoutePass.php]: at this moment, we can break the foreach loop.

**Comment by WouterJ** [comment]
The XML schema needs to be updated

File: DependencyInjection/Compiler/AutoRoutePass.php (hunk 3 above)
Comment (dantleech) [inline DependencyInjection/Compiler/AutoRoutePass.php]: Yes

**Comment by dantleech** [comment]
Also PHPCR-ODM tries to load the mapping for the PHPCR-ODM document, even with `auto_mapping: false`, and this fails because it requires RoutingBundle.

**Comment by dantleech** [comment]
So the issue with this PR is that I am allowing the user to choose the "adapter" which will be used. This is good.

However, this conflicts with the "persistence" configuration:

- The bundle requires that you choose at least one `persistence` system. 
- Our use case is in using the PHPCR-ODM event listener, but not the adapter.

So, we could keep the `adapter` choice specified here. If it is *not* specified, we will use the configured persistence layers default adapter.

Seems hacky though, but it would be a shame to change the persistence config as it a standard.

@dbu do you have any suggestions?

**Comment by dantleech** [comment]
As an example, something like the following would work:

````yaml
adapter: sulu_phpcr_adapter
adapters:
    doctrine_phpcr_odm:
        disable_mapping: true
        event_listener: true
````

File: Adapter/PhpcrOdmAdapter.php [hunk 4]
```diff
@@ -189,7 +189,7 @@ public function getReferringAutoRoutes($contentDocument)
     /**
      * {@inheritDoc}
      */
-    public function findRouteForUri($uri)
+    public function findRouteForUri($uri, UriContext $uri)
```
Comment (dantleech) [inline Adapter/PhpcrOdmAdapter.php]: Changes according to: https://github.com/symfony-cmf/RoutingAuto/pull/32

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (dantleech) [inline DependencyInjection/Configuration.php]: Even if we add an `->info()`?

Comment (WouterJ) [inline DependencyInjection/Configuration.php]: I think most of the time, when a node has only one method call (in this case `defaultValue()`), everything is put on one line. As soon as you have multiple method calls, it's indented like this.

At the moment, there is no `info()`, but if you add it, putting them on multiple lines would be better.

**Comment by dantleech** [comment]
ok .. 

- You can now specify an adapter via. the `cmf_routing_auto.adapter` config key.
- If an persistence system has been enabled the adapter will be implicitly set.
- Explicitly configured adapter will override implicit.

The Document mapping is only registered if the `RoutingBundle` is also registered.

**Comment by dantleech** [comment]
@dbu @WouterJ good with you guys?

**Comment by dbu** [comment]
seems ok to me, but did not review in all detail.

File: DependencyInjection/Compiler/AdapterPass.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class AdapterPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition(
+            'cmf_routing_auto.auto_route_manager'
```
Comment (dantleech) [inline DependencyInjection/Compiler/AdapterPass.php]: Why do we do these checks?

**Comment by dantleech** [comment]
@WouterJ will merge this this evening if you want to check it

**Comment by dantleech** [comment]
Merged

**Pull request merged**